AArch64: Implement _virtualUnresolvedHelper in PicBuilder.spp

knn-k opened the pull request:
This commit implements _virtualUnresolvedHelper in PicBuilder.spp for
AArch64.

Signed-off-by: knn-k <konno@jp.ibm.com>

knn-k commented:
This assumes the instruction sequence generated by `PrivateLinkage::buildVirtualDispatch()`, which is going to be implemented later.

PR #5432 adds the same `L_commonLookupException` as this commit does.

knn-k commented:
This commit uses `flushICache` added in PR #5246.  Marking as "WIP:" until that PR is merged.

knn-k commented:
This PR now depends on PR #5908 instead of #5246 as the flushICache change is separated.

knn-k commented:
Resolved merge conflict with #5246.

knn-k commented:
Corrected comment alignment.

0xdaryl commented on runtime/compiler/aarch64/runtime/PicBuilder.spp:
```diff
@@ -293,6 +363,12 @@ const_mcc_callPointPatching_unwrapper:
 	.align	3
 	.dword	mcc_callPointPatching_unwrapper
 
+L_commonLookupException:
```
This needs to be reconciled with the version in PR #5432.

knn-k replied:
Removed duplicated `L_commonLookupException`, and resolved merge conflicts.

0xdaryl commented on runtime/compiler/aarch64/runtime/PicBuilder.spp:
```diff
@@ -243,8 +251,70 @@ _interpreterUnresolvedInstanceDataGlue:
+	neg	x1, x0						// convert vTable offset to positive value
+	cmp	x1, #4095					// test if index is within the 12bit range
+	bgt	L_virtualUnresolvedHelperWideIndex		// If index > 4095
+	sub	x0, x2, #8					// get address of the ldr instruction
```
It looks like the following code is patching up an instruction sequence in the mainline code (i.e., not the virtual call snippet).  Since that code is not hooked up yet, can you describe what that instruction sequence looks like?  It will help understand what's going on here.  Thanks.

knn-k replied:
Please take a look at the comment lines above the label `_virtualUnresolvedHelper:` for the mainline code that is patched.

knn-k commented on runtime/compiler/aarch64/runtime/PicBuilder.spp:
```diff
@@ -85,6 +87,12 @@
 	.set	J9TR_USCSnippet_HelperOffset,	0xFF000000
 	.set	J9TR_USCSnippet_HelperOffsetShift,	24
 
+// Unresvoled virtual call snippet
```
Corrected spelling.

knn-k commented on runtime/compiler/aarch64/runtime/PicBuilder.spp:
```diff
@@ -243,8 +251,70 @@ _interpreterUnresolvedInstanceDataGlue:
 _interpreterUnresolvedInstanceDataStoreGlue:
 	hlt	#0	// Not implemented yet
 
+// For vitual unresolved call, we generate following instructions
```
Corrected spelling.

knn-k commented:
Some problems found, and I fixed them:
- `x9` is supposed to hold the vTable index, and it must not be destroyed.  Use `blr dstReg` instead of `blr x9` for calling the target method, allocating a register for that purpose.
- The jump target of the `ret` instruction at the end of `_virtualUnresolvedHelper` was wrong.  Adjusted it to point to the `movz` instruction.
- Stopped encoding the vTable index into the `ldr` instruction.  Use of negative offset in `ldr` is restricted in AArch64.
- No need to implement the separate `wideIndex` case.

knn-k commented:
Changed the use of `x28` in PicBuilder.spp, in parallel with #7243.

0xdaryl commented:
Jenkins test sanity alinux64 jdk11

0xdaryl commented:
Arraylet support (balanced GC) is not supported in AArch64 yet, hence the CI failures.
